src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs

```diff
@@ -293,67 +293,41 @@ private static void GenerateNodes(INamespaceSymbol globalNamespace, List<Node> l
             list.Add(node);
 
             // Add all child members
-            // Use GetMemberNames, because we only want the names for the symbols that INamedTypeSymbol.MemberNames returns
-            var memberNames = GetMemberNames(globalNamespace).Distinct();
             var memberLookup = GetMembers(globalNamespace).ToLookup(c => c.Name);
 
-            foreach (var memberName in memberNames)
+            foreach (var grouping in memberLookup)
             {
-                GenerateNodes(memberName, 0 /*index of root node*/, memberLookup[memberName], list);
+                GenerateNodes(grouping.Key, 0 /*index of root node*/, grouping, list);
             }
         }
 
+        private static readonly Func<ISymbol, bool> UseSymbol = 
+            s => s.CanBeReferencedByName && s.DeclaredAccessibility != Accessibility.Private;
+
         // generate nodes for symbols that share the same name, and all their descendants
-        private static void GenerateNodes(string name, int parentIndex, IEnumerable<INamespaceOrTypeSymbol> symbolsWithSameName, List<Node> list)
+        private static void GenerateNodes(string name, int parentIndex, IEnumerable<ISymbol> symbolsWithSameName, List<Node> list)
         {
             var node = new Node(name, parentIndex);
             var nodeIndex = list.Count;
             list.Add(node);
 
             // Add all child members
-            // Use GetMemberNames, because we only want the names for the symbols that INamedTypeSymbol.MemberNames returns
-            var memberNames = symbolsWithSameName.SelectMany(c => GetMemberNames(c)).Distinct();
-            var memberLookup = symbolsWithSameName.SelectMany(c => GetMembers(c)).ToLookup(c => c.Name);
+            var membersByName = symbolsWithSameName.SelectMany(GetMembers).ToLookup(s => s.Name);
 
-            foreach (var memberName in memberNames)
+            foreach (var grouping in membersByName)
             {
-                GenerateNodes(memberName, nodeIndex, memberLookup[memberName], list);
+                GenerateNodes(grouping.Key, nodeIndex, grouping, list);
             }
         }
 
-        private static IEnumerable<string> GetMemberNames(ISymbol symbol)
+        private static Func<ISymbol, IEnumerable<ISymbol>> GetMembers = symbol =>
```

Comment:
@gafter  I was under the impression that the compiler wasn't caching delegates for static methods *yet*.  As such, we would have a ton of delegates instances created as we recursively walked over all the symbols.  Because of that, I'm doing the manual caching to prevent that.

Is this unnecessary at this point?